processcsv.php

```diff
@@ -13,16 +13,94 @@
 // Setup a new form instance
 $mform = new oneroster_csv_form();
 
+// Function to recursively delete a directory
+function delete_directory($dir) {
+    if (!file_exists($dir)) {
+        return true;
+    }
+    if (!is_dir($dir)) {
+        return unlink($dir);
+    }
+    foreach (scandir($dir) as $item) {
+        if ($item == '.' || $item == '..') {
+            continue;
+        }
+        if (!delete_directory($dir . DIRECTORY_SEPARATOR . $item)) {
+            return false;
+        }
+    }
+    return rmdir($dir);
+}
+// Function to check the manifest.csv file and validate required files
+function check_manifest_and_files($manifest_path, $tempdir) {
+    $required_files = [];
+    if (($handle = fopen($manifest_path, "r")) !== FALSE) {
+        while (($data = fgetcsv($handle, 1000, ",")) !== FALSE) {
+            if (in_array($data[1], ['bulk', 'delta'])) {
+                $required_files[] = str_replace('file.', '', $data[0]) . '.csv';
```

Comment:
Please add a check for local file inclusion, eg dont allow traversal outside the directory.

Look into parameter cleaning in Moodle.

clean_param($myparameter, PARAM_FILE); // For data coming in.

format_text($mytext, FORMAT_MOODLE); // For potentially dodgy data being ouptut to the screen.